internal/ansible/metrics/metrics.go

```diff
@@ -82,6 +84,132 @@ func RegisterBuildInfo(r prometheus.Registerer) {
+	Summary   *UserMetricSummary   `json:"summary,omitempty" yaml:"summary,omitempty"`
+}
+
+type UserMetricCounter struct {
+	Inc bool    `json:"increment,omitempty" yaml:"increment,omitempty"`
+	Add float64 `json:"add,omitempty" yaml:"add,omitempty"`
+}
+
+type UserMetricGauge struct {
+	Set              float64 `json:"set,omitempty" yaml:"set,omitempty"`
+	Inc              bool    `json:"increment,omitempty" yaml:"increment,omitempty"`
+	Dec              bool    `json:"decrement,omitempty" yaml:"decrement,omitempty"`
+	Add              float64 `json:"add,omitempty" yaml:"add,omitempty"`
+	Sub              float64 `json:"subtract,omitempty" yaml:"subtract,omitempty"`
+	SetToCurrentTime bool    `json:"set_to_current_time,omitempty" yaml:"set_to_current_time,omitempty"`
+}
+
+type UserMetricHistogram struct {
+	Observe float64 `json:"observe,omitempty" yaml:"observe,omitempty"`
+}
+
+type UserMetricSummary struct {
+	Observe float64 `json:"observe,omitempty" yaml:"observe,omitempty"`
+}
+
+func HandleUserMetric(r prometheus.Registerer, metricSpec UserMetric) error {
+	defer recoverMetricPanic()
+	logf.Log.WithName("metrics").Info("Registering", "metric", metricSpec.Name)
+	_, ok := userMetrics[metricSpec.Name]
+	logf.Log.WithName("metrics").Info("at least this worked 1")
+	if ok != true {
+		// This is the first time we've seen this metric
+		// TODO(asmacdo) elif metricSpec.* foreach
+		logf.Log.WithName("metrics").Info("at least this worked 2")
+		if metricSpec.Counter != nil {
+			logf.Log.WithName("metrics").Info("at least this worked 3")
+			userMetrics[metricSpec.Name] = prometheus.NewCounter(prometheus.CounterOpts{
+				Name: metricSpec.Name,
+				Help: metricSpec.Help,
+			})
+		}
+		if metricSpec.Gauge != nil {
+			logf.Log.WithName("metrics").Info("at least this worked 4")
+			userMetrics[metricSpec.Name] = prometheus.NewGauge(prometheus.GaugeOpts{
+				Name: metricSpec.Name,
+				Help: metricSpec.Help,
+			})
+		}
+		if metricSpec.Histogram != nil {
+			userMetrics[metricSpec.Name] = prometheus.NewHistogram(prometheus.HistogramOpts{
+				Name: metricSpec.Name,
+				Help: metricSpec.Help,
+			})
+		}
+		if metricSpec.Summary != nil {
+			userMetrics[metricSpec.Name] = prometheus.NewSummary(prometheus.SummaryOpts{
+				Name: metricSpec.Name,
+				Help: metricSpec.Help,
+			})
+		}
+
+		// TODO(asmacdo) mustregister foreach
+		logf.Log.WithName("metrics").Info("at least this worked 5")
+		r.MustRegister(userMetrics[metricSpec.Name])
+	}
+	logf.Log.WithName("metrics").Info("Metric found")
+	logf.Log.WithName("metrics").Info("Metric found", "Name:", metricSpec.Name)
+
+	collector := userMetrics[metricSpec.Name]
+
+	logf.Log.WithName("metrics").Info("about to try the switch")
+	logf.Log.WithName("metrics").Info(fmt.Sprintf("%v", collector))
+	switch v := collector.(type) {
+	// TODO(asmacdo) remove this
+	case nil:
+		logf.Log.WithName("metrics").Info("Well this certainly is a problem")
```

Comment:
Should this be an error?

Reply:
This can't happen, just a debug sanity check